java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java

```diff
@@ -338,14 +338,20 @@ public void putNext() {
     @Override
     public void putNext(ArrowBuf appMetadata) {
       ArrowRecordBatch batch = unloader.getRecordBatch();
-      while (!observer.isReady()) {
+      // Check isCancelled as well to avoid inadvertently blocking forever
+      // (so long as PutListener properly implements it)
+      while (!observer.isReady() && !listener.isCancelled()) {
         /* busy wait */
       }
-      try {
-        // Takes ownership of appMetadata
-        observer.onNext(new ArrowMessage(batch, appMetadata));
+      // ArrowMessage takes ownership of appMetadata and batch
+      // gRPC should take ownership of ArrowMessage, but in some cases it doesn't, so guard against it
+      // ArrowMessage#close is a no-op if gRPC did its job
+      try (final ArrowMessage message = new ArrowMessage(batch, appMetadata)) {
+        observer.onNext(message);
       } catch (StatusRuntimeException sre) {
         throw StatusUtils.fromGrpcRuntimeException(sre);
+      } catch (Exception e) {
+        throw StatusUtils.fromThrowable(e);
```

Comment:
It seems StatusUtils.fromThrowable will call fromGrpcRuntimeException internally, so we only need one catch clause here?